pkg/storage/stores/tsdb/single_file_index.go

```diff
@@ -118,26 +125,61 @@ func (f *TSDBFile) Reader() (io.ReadSeeker, error) {
 // and translates the IndexReader to an Index implementation
 // It loads the file into memory and doesn't keep a file descriptor open
 type TSDBIndex struct {
-	reader      IndexReader
-	chunkFilter chunk.RequestChunkFilterer
+	reader         IndexReader
+	chunkFilter    chunk.RequestChunkFilterer
+	postingsReader PostingsReader
 }
 
 // Return the index as well as the underlying raw file reader which isn't exposed as an index
 // method but is helpful for building an io.reader for the index shipper
-func NewTSDBIndexFromFile(location string) (*TSDBIndex, GetRawFileReaderFunc, error) {
+func NewTSDBIndexFromFile(location string, opts IndexOpts) (Index, GetRawFileReaderFunc, error) {
 	reader, err := index.NewFileReader(location)
 	if err != nil {
 		return nil, nil, err
 	}
 
-	return NewTSDBIndex(reader), func() (io.ReadSeeker, error) {
+	postingsReader := getPostingsReader(reader, opts.PostingsCache)
+	tsdbIdx := NewTSDBIndex(reader, postingsReader)
+
+	return tsdbIdx, func() (io.ReadSeeker, error) {
 		return reader.RawFileReader()
 	}, nil
 }
 
-func NewTSDBIndex(reader IndexReader) *TSDBIndex {
+func getPostingsReader(reader IndexReader, postingsCache cache.Cache) PostingsReader {
+	var pr PostingsReader
+
+	if postingsCache != nil {
+		pr = NewCachedPostingsReader(reader, util_log.Logger, postingsCache)
+	}
+
+	if pr == nil {
+		pr = NewPostingsReader(reader)
+	}
+
+	return pr
+}
+
+func NewPostingsReader(reader IndexReader) PostingsReader {
+	return &simplePostingsReader{reader: reader}
+}
+
+type simplePostingsReader struct {
```

Comment:
nit: Personally I don't like to use the word "simple" for objects. I'd use "default", since it's the default implementation of the `PostingsReader` interface.

```suggestion
type defaultPostingsReader struct {
```